ddtrace/contrib/wsgi/wsgi.py

```diff
@@ -80,73 +94,140 @@ def default_wsgi_span_modifier(span, environ):
 
+            with self.tracer.trace(self.response_span) as resp_span:
+                self.response_span_modifier(resp_span, result)
                 for chunk in result:
                     yield chunk
 
-            url = construct_url(environ)
-            method = environ.get("REQUEST_METHOD")
-            query_string = environ.get("QUERY_STRING")
-            request_headers = get_request_headers(environ)
-            trace_utils.set_http_meta(
-                span, config.wsgi, method=method, url=url, query=query_string, request_headers=request_headers
-            )
-
-            if self.span_modifier:
-                self.span_modifier(span, environ)
-
             if hasattr(result, "close"):
                 try:
                     result.close()
                 except Exception:
                     typ, val, tb = sys.exc_info()
-                    span.set_exc_info(typ, val, tb)
+                    req_span.set_exc_info(typ, val, tb)
                     six.reraise(typ, val, tb=tb)
+
+    def traced_start_response(self, start_response, request_span, status, environ, exc_info=None):
+        # type: (Callable, Span, str, Dict, Any) -> None
+        """sets the status code on a request span when start_response is called"""
+        status_code, _ = status.split(" ", 1)
+        trace_utils.set_http_meta(request_span, self.config, status_code=status_code)
+        return start_response(status, environ, exc_info)
+
+    def request_span_modifier(self, req_span, environ):
+        # type: (Span, Dict) -> None
+        """Implement to modify span attributes on the request_span"""
+        pass
+
+    def application_span_modifier(self, app_span, environ, result):
+        # type: (Span, Dict, Iterable) -> None
+        """Implement to modify span attributes on the application_span"""
+        pass
+
+    def response_span_modifier(self, resp_span, response):
+        # type: (Span, Dict) -> None
+        """Implement to modify span attributes on the request_span"""
+        pass
+
+
+class DDWSGIMiddleware(DDWSGIMiddlewareBase):
+    """WSGI middleware providing tracing around an application.
+
+    :param application: The WSGI application to apply the middleware to.
+    :param tracer: Tracer instance to use the middleware with. Defaults to the global tracer.
+    :param span_modifier: Span modifier that can add tags to the root span.
+                            Defaults to using the request method and url in the resource.
+    """
+
+    request_span = "wsgi.request"
+    application_span = "wsgi.application"
+    response_span = "wsgi.response"
+
+    def __init__(self, application, tracer=None, span_modifier=default_wsgi_span_modifier):
+        # type: (Iterable, Tracer, Callable) -> None
+        super(DDWSGIMiddleware, self).__init__(application, tracer or ddtrace.tracer, config.wsgi, None)
+        # Deprecate - This callable was used used to modify wsgi.request spans.
```

Comment:
I am not clear on why this deprecation is added here. Overriding the `DDWSGIMiddleware.request_span_modifier` is certainly one option. But the current interface is such that a user can use the wsgi middleware and include additional modifications to the request span than what is provided by the wsgi middleware. If we were to add another method for supporting such customizations, then we could deprecate the current interface, but that's not what is done here.

Reply:
We can discuss this later. I'll resolve this conversation since this issue does not affect these changes.